# 부산시 건물용도 코드파일

In [56]:
import geopandas as gpd
import pandas as pd
import os

## 데이터 불러오기

- gpd_loader.py에서 get_gpd함수 가져오기
- 부산 내 16개의 구의 건물용도 .shp파일을 불러서 df로 저장하는 함수

In [ ]:
from gpd_loader import get_gpd

dir_path = '../../../Data_list/source_data/용도별건물공간확보_부산' 
자치구_list, df_dict = get_gpd(dir_path)

- 각 구이름을 key로 df는 value에 저장되어있음

In [36]:
df_dict = dict(sorted(df_dict.items()))

In [37]:
df_dict.keys()

dict_keys(['강서구', '금정구', '기장군', '남구', '동구', '동래구', '부산진구', '북구', '사상구', '사하구', '서구', '수영구', '연제구', '영도구', '중구', '해운대구'])

In [38]:
df_dict

{'강서구':              A0                            A1                   A2  \
 0      32511418  2002199031551910905200000000  2644010100101940001   
 1      32491218  2002198998511910971000000000  2644010100101940008   
 2      32553288  2002198987001910982000000000  2644010100101940009   
 3      28657236  2006199054231910081000000000  2644010100101950004   
 4      32553289  1940199003591908491200000000  2644010100101990054   
 ...         ...                           ...                  ...   
 24414  43618957  2006188888261763736800000000  2644011700102160010   
 24415  29443316  2005188800211763684800000000  2644011700102170000   
 24416  54926723  2004188792421762733000000000  2644011700102180000   
 24417  29443317  2003188792831762503700000000  2644011700102180013   
 24418  32250703  2004188962001761900600000000  2644011700102200004   
 
                A3              A4 A5  A6      A7          A8 A9  ...     A27  \
 0      2644010100  부산광역시 강서구 대저1동  1  일반   194-1   103916

- 원본 유지를 위해 복사본 생성

In [39]:
df_dict_copy = df_dict.copy() 

In [40]:
df_dict_copy['강서구'].head(3)

,A0,A1,A2,A3,A4,A5,A6,A7,A8,A9,...,A27,A28,A29,A30,A31,A32,A33,A34,A35,geometry
0,32511418,2002199031551910905200000000,2644010100101940001,2644010100,부산광역시 강서구 대저1동,1,일반,194-1,103916038,1,...,축사,3,농수산용,5.4,1.0,0.0,2002-03-28,2002-06-04,2025-01-10 02:22:27.4122,"POLYGON ((381208.163 293218.43, 381204.375 293..."
1,32491218,2002198998511910971000000000,2644010100101940008,2644010100,부산광역시 강서구 대저1동,1,일반,194-8,103916039,1,...,기타창고시설,7,기타,5.4,1.0,NaN,2002-03-28,2002-06-04,2025-01-10 02:22:27.4122,"POLYGON ((381143.488 293230.135, 381173.095 29..."
2,32553288,2002198987001910982000000000,2644010100101940009,2644010100,부산광역시 강서구 대저1동,1,일반,194-9,103916040,1,...,사무소,2,상업용,5.4,1.0,NaN,2002-03-28,2002-06-04,2025-01-10 02:22:27.4122,"POLYGON ((381144.389 293228.515, 381149.731 29..."


## EDA

- A25:주요용도명
- A27:세부용도명
- A29:건물용도명 

In [41]:
columns = ['A25' , 'A27', 'A29']

In [42]:
df_dict['강서구'][columns].head()

,A25,A27,A29
0,동.식물 관련시설,축사,농수산용
1,창고시설,기타창고시설,기타
2,제2종근린생활시설,사무소,상업용
3,제2종근린생활시설,제조업소,상업용
4,단독주택,단독주택,주거용


- 강서구를 예시로 A29, A25, A27순으로 더 세세하게 나뉨
- 그 값들을 각각 확인한다

In [43]:
print('건물용도_종류 \n: ', df_dict_copy['강서구'].A29.unique(), '\n')
print('주요용도_종류 \n: ', df_dict_copy['강서구'].A25.unique(), '\n')
print('세부용도_종류 \n: ', df_dict_copy['강서구'].A27.unique(), '\n')
print(len(df_dict_copy['강서구'].A29.unique()),'개' , len(df_dict_copy['강서구'].A25.unique()), '개', len(df_dict_copy['강서구'].A27.unique()),'개')

건물용도_종류 
:  ['농수산용' '기타' '상업용' '주거용' '문교사회용' '공업용' None '공공용'] 

주요용도_종류 
:  ['동.식물 관련시설' '창고시설' '제2종근린생활시설' '단독주택' '근린생활시설' '제1종근린생활시설' '분뇨.쓰레기처리시설'
 '교육연구및복지시설' '공장' '교육연구시설' '노유자시설' '문화및집회시설' '업무시설' '위험물저장및처리시설' None
 '공공용시설' '판매시설' '판매및영업시설' '종교시설' '운수시설' '운동시설' '숙박시설' '자동차관련시설' '의료시설'
 '공동주택' '방송통신시설' '관광휴게시설' '교정및군사시설' '제조업소' '위락시설'] 

세부용도_종류 
:  ['축사' '기타창고시설' '사무소' '제조업소' '단독주택' '창고' '근린생활시설' '기타사무소' '기타제2종근생활시설'
 '수리점' '제2종근린생활시설' '제1종근린생활시설' '소매점' '일반음식점' '창고시설' '일반창고' '양수장'
 '기타동식물관련시설' '분뇨.쓰레기처리시설' '교육연구및복지시설' '마을공동시설' '표구점' '일반공장' '기타학교'
 '사회복지시설' '학원' '문화및집회시설' '버섯재배사' '기타일반업무시설' '기타공공시설' '업무시설' '기타위험물저장처리시설'
 '주유소' '분뇨처리시설' '휴게음식점' '다가구주택' '마을회관' None '동.식물 관련시설' '공공용시설' '일반목욕장'
 '이(미)용원' '기타제1종근생활시설' '부동산중개사무소' '기타교육연구및복지시설' '기타판매시설' '교회' '의약품도매점'
 '액화석유가스충전소' '위험물저장소' '마을공동구판장' '의원' '기타근린생활시설' '기타종교집회장' '판매및영업시설' '성당'
 '공공시설' '부화장' '도축장' '노래연습장' '고등학교' '우체국' '영유아보육시설' '유치원' '공항시설' '부동산중개업소'
 '초등학교' '승마장' '작물재배사' '기타공공용시설' '기타공공업무시설' '노인복지시설' '보건소' '공장' '자동차영업소'
 '석유판매

> 구 이름을 변경하면서 살펴본 결과 각 구마다 건물용도 종류가 다름 

- 건물용도에 해당하는 열(A25, A27, A29)에 해당하는 열만 남기고
    - 3가지 모두 None인 건물은 용도를 알 수 없으므로 drop
    - 가장 세분화된 열인 A27을 메인으로 사용함
        - 세부용도명(A27)이 None일 시, 주요용도명(A25)의 값으로 채움
        - 세부용도명(A27)과 주요용도명(A25) 모두 None일 시, 건물용도명(A29)의 값으로 채움

## 전처리

In [44]:
def fill_na(df):
    df = df.dropna(subset=columns, how='all')
    df.loc[df['A27'].isna(), 'A27'] = df.loc[df['A27'].isna(), 'A25']
    df.loc[df['A27'].isna(), 'A27'] = df.loc[df['A27'].isna(), 'A29']
    return df

- 각 구마다 적용하기 위해 for문을 사용
    - fillna()함수를 적용
    - A27, A25, A29 컬럼만 남김

In [45]:
for df_key, df in df_dict_copy.items():
    df_dict_copy[df_key] = fill_na(df)
    df_dict_copy[df_key] = df_dict_copy[df_key][columns]
    

- 각 구의 df에서 A27에 있는 값들을 전부 뽑아서 list화 하는데 set을 이용하여 중복을 제거

In [46]:
a27_list = list(set(val for df in df_dict_copy.values() for val in df['A27'].unique()))

- 16개 구에 해당하는 건물세부용도 이름이 전부 a27_list에 저장됐고 이를 살펴보기위해 txt파일로 변환

In [ ]:
with open('../../../Data_list/preprocessing_result/건물공간_데이터/a27_list.txt', 'w', encoding='utf-8') as f:
    for item in a27_list:
        f.write(f"{item}\n")

In [48]:
len(a27_list)

352

- 352개의 건물세부용도명을 살펴본 결과
- 학교(대학교, 고등학교 제외), 유치원, 어린이집, 노유자시설(어린이 시설이 포함되었기에), 주거용 건물 총 19개로 추려짐
- 이를 nearby_list에 저장
- 건물 선정 이유는 학교, 주거용 근처에 들락날락이 존재해야 부모님을 기다리는 시간에 편하게 방문하기 좋기 때문

In [ ]:
with open('../../../Data_list/preprocessing_result/건물공간_데이터/부산시_주거지및학교_list.txt', 'r', encoding='utf-8') as f:
    부산시_주거지및학교_list = [line.strip() for line in f.readlines()]

In [50]:
부산시_주거지및학교_list

['영유아보육시설',
 '다세대주택',
 '아동복지시설',
 '단독주택',
 '초등학교',
 '유치원',
 '주거용',
 '어린이집',
 '아파트',
 '학원',
 '중학교',
 '지역아동센터',
 '기타아동관련시설',
 '다가구주택',
 '공동주택',
 '노유자시설',
 '다중주택',
 '아동관련시설',
 '학교']

- 원본 데이터프레임에 앞서 언급한 fill_na함수를 적용 (copy본은 A25, A27, A29열만 뽑아놨으니 쓸 수 없음)
- 데이터 프레임에 A27의 값이 nearby_list에 있는 행만 남김
- 각 구별 남은 건물용도명과 어떤것들이 있는지 확인

In [51]:
for key in df_dict.keys():
    df_dict[key] = fill_na(df_dict[key])
    df_dict[key] = df_dict[key][df_dict[key]['A27'].isin(부산시_주거지및학교_list)]
    print(f'{key}의 A27 값 {len(df_dict[key].A27.unique())} \n: ',df_dict[key].A27.unique())

강서구의 A27 값 12 
:  ['단독주택' '학원' '다가구주택' '영유아보육시설' '유치원' '초등학교' '다중주택' '주거용' '어린이집' '중학교'
 '다세대주택' '아파트']
금정구의 A27 값 17 
:  ['단독주택' '다가구주택' '학원' '기타아동관련시설' '아동복지시설' '초등학교' '다중주택' '어린이집' '영유아보육시설'
 '유치원' '중학교' '지역아동센터' '다세대주택' '노유자시설' '공동주택' '주거용' '아파트']
기장군의 A27 값 16 
:  ['단독주택' '다가구주택' '학원' '기타아동관련시설' '영유아보육시설' '초등학교' '아동복지시설' '공동주택' '어린이집'
 '지역아동센터' '유치원' '중학교' '다중주택' '다세대주택' '노유자시설' '아파트']
남구의 A27 값 15 
:  ['다가구주택' '단독주택' '학원' '공동주택' '초등학교' '유치원' '중학교' '영유아보육시설' '어린이집' '기타아동관련시설'
 '노유자시설' '다중주택' '아동복지시설' '다세대주택' '아파트']
동구의 A27 값 16 
:  ['단독주택' '학원' '다가구주택' '지역아동센터' '유치원' '초등학교' '중학교' '영유아보육시설' '다중주택' '공동주택'
 '어린이집' '기타아동관련시설' '아동관련시설' '다세대주택' '노유자시설' '아파트']
동래구의 A27 값 16 
:  ['단독주택' '다가구주택' '기타아동관련시설' '어린이집' '초등학교' '중학교' '학원' '영유아보육시설' '유치원' '공동주택'
 '노유자시설' '지역아동센터' '다중주택' '다세대주택' '아동복지시설' '아파트']
부산진구의 A27 값 17 
:  ['단독주택' '다가구주택' '학원' '초등학교' '중학교' '유치원' '기타아동관련시설' '어린이집' '아동복지시설'
 '영유아보육시설' '노유자시설' '다세대주택' '아동관련시설' '공동주택' '지역아동센터' '다중주택' '아파트']
북구의 A27 값 15 
:  ['학원' '단독주택' '다가구주택' '

In [52]:
a = 0
for key in df_dict.keys():
    print(len(df_dict[key]))
    a += len(df_dict[key])
print(a)

9589
15922
8864
13782
9862
12808
17549
8220
9316
16436
9608
10417
9889
10325
3750
13921
180258


## 수정된 데이터 저장

- 각 구마다 데이터프레임을 .shp파일로 저장

In [ ]:
os.makedirs('../../../Data_list/preprocessing_result/건물공간_데이터/용도별건물공간확보_부산_가공', exist_ok=True)
for name in df_dict.keys():
    os.makedirs(f'../../../Data_list/preprocessing_result/건물공간_데이터/용도별건물공간확보_부산_가공/{name}', exist_ok=True)
    df_dict[name].to_file(f'../../../Data_list/preprocessing_result/건물공간_데이터/용도별건물공간확보_부산_가공/{name}/{name}.shp', encoding='cp949')

- 각 데이터프레임에 '자치구'라는 이름의 열을 추가
    - '자치구'열에 자치구 이름을 값으로 추가


In [54]:
for name, df in df_dict.items():
    df['자치구'] = name 

- 16개의 데이터프레임을 하나의 데이터프레임으로 합침

In [57]:
df_merge = pd.concat(df_dict.values(), ignore_index=True)

- 데이터프레임의 정보를 살펴보니 위에서 각 데이터프레임의 길이를 합침 값 a와 동일한 것으로 확인

In [58]:
len(df_merge)

180258

In [59]:
df_merge.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 180258 entries, 0 to 180257
Data columns (total 38 columns):
 #   Column    Non-Null Count   Dtype   
---  ------    --------------   -----   
 0   A0        180258 non-null  int32   
 1   A1        180258 non-null  object  
 2   A2        180258 non-null  object  
 3   A3        180258 non-null  object  
 4   A4        180258 non-null  object  
 5   A5        180258 non-null  object  
 6   A6        180258 non-null  object  
 7   A7        180255 non-null  object  
 8   A8        180258 non-null  object  
 9   A9        180258 non-null  object  
 10  A10       180258 non-null  object  
 11  A11       180258 non-null  object  
 12  A12       180258 non-null  object  
 13  A13       27519 non-null   object  
 14  A14       16553 non-null   object  
 15  A15       180254 non-null  object  
 16  A16       180254 non-null  object  
 17  A17       177268 non-null  float64 
 18  A18       180000 non-null  float64 
 19  A19       18025

- 종합한 데이터프레임을 csv파일로 저장

In [ ]:
df_merge.to_csv('../../../Data_list/preprocessing_result/건물공간_데이터/부산시_주거및학교건물데이터.csv')

## 유아시설 건물만 따로 뽑기

- 위에서 주거+학교 데이터를 뽑았음
- 이후에 유아시설만 있는 데이터가 따로 필요해 추가 가공

In [ ]:
with open('../../../Data_list/preprocessing_result/건물공간_데이터/부산_유아시설_list.txt', 'r', encoding='utf-8') as f:
    부산시_유아시설_list = [line.strip() for line in f.readlines()]

In [62]:
부산시_유아시설_list

['영유아보육시설', '아동복지시설', '초등학교', '유치원', '어린이집', '지역아동센터', '기타아동관련시설', '아동관련시설']

In [63]:
for key in df_dict.keys():
    df_dict[key] = fill_na(df_dict[key])
    df_dict[key] = df_dict[key][df_dict[key]['A27'].isin(부산시_유아시설_list)]
    print(f'{key}의 A27 값 {len(df_dict[key].A27.unique())} \n: ',df_dict[key].A27.unique())

강서구의 A27 값 4 
:  ['영유아보육시설' '유치원' '초등학교' '어린이집']
금정구의 A27 값 7 
:  ['기타아동관련시설' '아동복지시설' '초등학교' '어린이집' '영유아보육시설' '유치원' '지역아동센터']
기장군의 A27 값 7 
:  ['기타아동관련시설' '영유아보육시설' '초등학교' '아동복지시설' '어린이집' '지역아동센터' '유치원']
남구의 A27 값 6 
:  ['초등학교' '유치원' '영유아보육시설' '어린이집' '기타아동관련시설' '아동복지시설']
동구의 A27 값 7 
:  ['지역아동센터' '유치원' '초등학교' '영유아보육시설' '어린이집' '기타아동관련시설' '아동관련시설']
동래구의 A27 값 7 
:  ['기타아동관련시설' '어린이집' '초등학교' '영유아보육시설' '유치원' '지역아동센터' '아동복지시설']
부산진구의 A27 값 8 
:  ['초등학교' '유치원' '기타아동관련시설' '어린이집' '아동복지시설' '영유아보육시설' '아동관련시설' '지역아동센터']
북구의 A27 값 6 
:  ['초등학교' '유치원' '영유아보육시설' '어린이집' '기타아동관련시설' '지역아동센터']
사상구의 A27 값 7 
:  ['유치원' '초등학교' '영유아보육시설' '어린이집' '기타아동관련시설' '아동복지시설' '지역아동센터']
사하구의 A27 값 7 
:  ['초등학교' '어린이집' '영유아보육시설' '기타아동관련시설' '유치원' '아동복지시설' '지역아동센터']
서구의 A27 값 7 
:  ['유치원' '초등학교' '어린이집' '지역아동센터' '영유아보육시설' '기타아동관련시설' '아동복지시설']
수영구의 A27 값 6 
:  ['유치원' '기타아동관련시설' '초등학교' '영유아보육시설' '어린이집' '아동복지시설']
연제구의 A27 값 7 
:  ['기타아동관련시설' '초등학교' '어린이집' '영유아보육시설' '유치원' '아동복지시설' '아동관련시설']
영도구의 A27 값 7 
:  ['초등학교' '영유아보육시설' '

In [64]:
a = 0
for key in df_dict.keys():
    print(len(df_dict[key]))
    a += len(df_dict[key])
print(a)

55
82
98
106
56
101
102
111
70
153
58
52
62
78
11
130
1325


In [ ]:
os.makedirs('../../../Data_list/preprocessing_result/건물공간_데이터/용도별건물공간확보_부산_가공', exist_ok=True)

In [66]:
for name, df in df_dict.items():
    df['자치구'] = name 

In [67]:
df_merge2 = pd.concat(df_dict.values(), ignore_index=True)

In [68]:
len(df_merge2)

1325

In [69]:
gdf = gpd.GeoDataFrame(df_merge2, geometry=df_merge2['geometry'])

In [ ]:
os.makedirs('../../../Data_list/preprocessing_result/건물공간_데이터/부산시_유아시설', exist_ok=True)

In [ ]:
gdf.to_file('../../../Data_list/preprocessing_result/건물공간_데이터/부산시_유아시설/부산시_유아시설_gdf.shp', encoding='cp949')